File: src/trade_bot/trading/reconciliation_diagnostics.py
"""Opt-in, deterministic diagnostics for paper reconciliation.

The recorder deliberately stores counters and timestamps only.  It never stores
payloads, credentials, or order identifiers and is inert unless enabled.
"""
from __future__ import annotations

from collections import Counter
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, Callable, Iterable


@dataclass
class ReconciliationDiagnostics:
    enabled: bool = False
    clock: Callable[[], datetime] = lambda: datetime.now(timezone.utc)
    selected_symbols: list[str] = field(default_factory=list)
    fetches: dict[str, dict[str, Any]] = field(default_factory=dict)
    gate_outcomes: Counter = field(default_factory=Counter)
    blockers: Counter = field(default_factory=Counter)
    signals_evaluated: int = 0
    signals_generated: int = 0
    paper_intents: int = 0
    fills: int = 0

    @staticmethod
    def data_age_seconds(observed_at: datetime | None, now: datetime) -> int | None:
        """Return non-negative age in seconds, or ``None`` for invalid data."""
        if observed_at is None:
            return None
        if observed_at.tzinfo is None:
            observed_at = observed_at.replace(tzinfo=timezone.utc)
        if now.tzinfo is None:
            now = now.replace(tzinfo=timezone.utc)
        age = int((now - observed_at).total_seconds())
        return age if age >= 0 else None

    def reset(self, selected_symbols: Iterable[str] = ()) -> None:
        if not self.enabled:
            return
        self.selected_symbols = list(selected_symbols)
        self.fetches.clear()
        self.gate_outcomes.clear()
        self.blockers.clear()
        self.signals_evaluated = self.signals_generated = 0
        self.paper_intents = self.fills = 0

    def _fetch(self, symbol: str) -> dict[str, Any]:
        return self.fetches.setdefault(symbol, {"attempts": 0, "successes": 0, "failures": 0})

    def record_fetch_attempt(self, symbol: str) -> None:
        if self.enabled:
            stats = self._fetch(symbol)
            stats["attempts"] += 1
            stats["last_attempt_at"] = self.clock()

    def record_fetch_result(self, symbol: str, success: bool, observed_at: datetime | None = None) -> None:
        if not self.enabled:
            return
        stats = self._fetch(symbol)
        stats["successes" if success else "failures"] += 1
        if success:
            observed_at = observed_at or self.clock()
            stats["observed_at"] = observed_at
            stats["last_success_at"] = observed_at

    def record_signal(self, generated: bool) -> None:
        if self.enabled:
            self.signals_evaluated += 1
            self.signals_generated += int(generated)

    def record_gate(self, outcome: str | None) -> None:
        if self.enabled:
            self.gate_outcomes[outcome or "unknown"] += 1

    def record_blocker(self, reason: str | None) -> None:
        if self.enabled:
            self.blockers[reason or "unknown"] += 1

    def record_paper_intent(self) -> None:
        if self.enabled:
            self.paper_intents += 1

    def record_fill(self) -> None:
        if self.enabled:
            self.fills += 1

    @staticmethod
    def _timestamp(value: datetime | None) -> str | None:
        return value.isoformat() if value is not None else None

    def snapshot(self, now: datetime | None = None) -> dict[str, Any]:
        if not self.enabled:
            return {"enabled": False}
        now = now or self.clock()
        fetches: dict[str, Any] = {}
        for symbol, stats in sorted(self.fetches.items()):
            fetches[symbol] = {
                "attempts": stats["attempts"],
                "successes": stats["successes"],
                "failures": stats["failures"],
                "data_age_seconds": self.data_age_seconds(stats.get("observed_at"), now),
                "last_attempt_at": self._timestamp(stats.get("last_attempt_at")),
                "last_success_at": self._timestamp(stats.get("last_success_at")),
            }
        return {
            "enabled": True,
            "selected_symbols": list(self.selected_symbols),
            "selected_symbol_count": len(self.selected_symbols),
            "signals_evaluated": self.signals_evaluated,
            "signals_generated": self.signals_generated,
            "paper_intents": self.paper_intents,
            "fills": self.fills,
            "fetches": fetches,
            "gate_outcomes": dict(sorted(self.gate_outcomes.items())),
            "blockers": dict(sorted(self.blockers.items())),
        }

File: src/trading/ReconciliationDiagnostics.cpp
#include "trading/ReconciliationDiagnostics.hpp"

#include <algorithm>

namespace trade::trading {

void ReconciliationDiagnostics::setEnabled(const bool enabled) {
  std::lock_guard<std::mutex> lock(mutex_);
  enabled_ = enabled;
}

DataAge ReconciliationDiagnostics::dataAge(const long long observed,
                                           const long long now) {
  if (observed <= 0 || now < observed) {
    return {};
  }
  return DataAge{true, now - observed};
}

void ReconciliationDiagnostics::reset(const std::vector<std::string> &selected_symbols) {
  if (!enabled_) {
    return;
  }
  std::lock_guard<std::mutex> lock(mutex_);
  selected_symbols_ = selected_symbols;
  fetches_.clear();
  gate_outcomes_.clear();
  blockers_.clear();
  signals_evaluated_ = 0;
  signals_generated_ = 0;
  paper_intents_ = 0;
  fills_ = 0;
}

void ReconciliationDiagnostics::recordFetchAttempt(const std::string &symbol) {
  if (!enabled_) return;
  std::lock_guard<std::mutex> lock(mutex_);
  ++fetches_[symbol].attempts;
}

void ReconciliationDiagnostics::recordFetchResult(const std::string &symbol,
                                                  const bool success,
                                                  const long long observed) {
  if (!enabled_) return;
  std::lock_guard<std::mutex> lock(mutex_);
  auto &stats = fetches_[symbol];
  if (success) {
    ++stats.successes;
    stats.last_success_epoch_seconds = observed;
  } else {
    ++stats.failures;
  }
}

void ReconciliationDiagnostics::recordSignal(const bool generated) {
  if (!enabled_) return;
  std::lock_guard<std::mutex> lock(mutex_);
  ++signals_evaluated_;
  if (generated) ++signals_generated_;
}

void ReconciliationDiagnostics::recordGateOutcome(const std::string &outcome) {
  if (!enabled_) return;
  std::lock_guard<std::mutex> lock(mutex_);
  ++gate_outcomes_[outcome.empty() ? "unknown" : outcome];
}

void ReconciliationDiagnostics::recordBlocker(const std::string &reason) {
  if (!enabled_) return;
  std::lock_guard<std::mutex> lock(mutex_);
  ++blockers_[reason.empty() ? "unknown" : reason];
}

void ReconciliationDiagnostics::recordPaperIntent() {
  if (!enabled_) return;
  std::lock_guard<std::mutex> lock(mutex_);
  ++paper_intents_;
}

void ReconciliationDiagnostics::recordFill() {
  if (!enabled_) return;
  std::lock_guard<std::mutex> lock(mutex_);
  ++fills_;
}

Json::Value ReconciliationDiagnostics::toJson(const long long now) const {
  Json::Value out(Json::objectValue);
  if (!enabled_) {
    out["enabled"] = false;
    return out;
  }
  std::lock_guard<std::mutex> lock(mutex_);
  out["enabled"] = true;
  out["selected_symbols"] = Json::arrayValue;
  for (const auto &symbol : selected_symbols_) out["selected_symbols"].append(symbol);
  out["selected_symbol_count"] = static_cast<Json::UInt64>(selected_symbols_.size());
  out["signals_evaluated"] = static_cast<Json::UInt64>(signals_evaluated_);
  out["signals_generated"] = static_cast<Json::UInt64>(signals_generated_);
  out["paper_intents"] = static_cast<Json::UInt64>(paper_intents_);
  out["fills"] = static_cast<Json::UInt64>(fills_);

  Json::Value fetches(Json::objectValue);
  for (const auto &[symbol, stats] : fetches_) {
    Json::Value item(Json::objectValue);
    item["attempts"] = static_cast<Json::UInt64>(stats.attempts);
    item["successes"] = static_cast<Json::UInt64>(stats.successes);
    item["failures"] = static_cast<Json::UInt64>(stats.failures);
    const DataAge age = dataAge(stats.last_success_epoch_seconds, now);
    item["data_age_seconds"] = age.available ? Json::Value(static_cast<Json::Int64>(age.seconds))
                                              : Json::Value(Json::nullValue);
    fetches[symbol] = item;
  }
  out["fetches"] = fetches;

  const auto mapToJson = [](const std::map<std::string, std::size_t> &values) {
    Json::Value result(Json::objectValue);
    for (const auto &[key, value] : values) result[key] = static_cast<Json::UInt64>(value);
    return result;
  };
  out["gate_outcomes"] = mapToJson(gate_outcomes_);
  out["blockers"] = mapToJson(blockers_);
  return out;
}

} // namespace trade::trading
